File: gradle_plugin_demo/custom_plugin/src/main/java/com/znh/gradle/custom/plugin/PageClassVisitor.java
package com.znh.gradle.custom.plugin;


import org.objectweb.asm.ClassVisitor;
import org.objectweb.asm.MethodVisitor;
import org.objectweb.asm.Opcodes;

/**
 * Created by znh on 2020-04-21
 * <p>
 * ASM操作class的类
 */
public class PageClassVisitor extends ClassVisitor {

    //当前类的类名称
    //本例：com/znh/gradle/plugin/demo/MainActivity
    private String className;

    //className类的父类名称
    //本例：androidx/appcompat/app/AppCompatActivity
    private String superName;

    public PageClassVisitor(ClassVisitor classVisitor) {
        super(Opcodes.ASM5, classVisitor);
    }

    @Override
    public void visit(int version, int access, String className, String signature, String superName, String[] interfaces) {
        super.visit(version, access, className, signature, superName, interfaces);
        this.className = className;
        this.superName = superName;
    }

    @Override
    public MethodVisitor visitMethod(int access, String methodName, String descriptor, String signature, String[] exceptions) {

        MethodVisitor methodVisitor = cv.visitMethod(access, methodName, descriptor, signature, exceptions);

        //判断是否自己要匹配的方法(这里匹配onCreate和onDestroy方法)
        if ("android/support/v7/app/AppCompatActivity".equals(superName)//support包
                || "androidx/appcompat/app/AppCompatActivity".equals(superName)) {//androidx包
            if (methodName.startsWith("onCreate")) {
                return new PageMethodVisitor(methodVisitor, className, methodName);
            }
            if (methodName.startsWith("onDestroy")) {
                return new PageMethodVisitor(methodVisitor, className, methodName);
            }
        }
        return methodVisitor;
    }
}

File: gradle80-plugin-demo/hui_plugin/src/main/java/com/znh/plugin/page/PageClassVisitor.kt
package com.znh.plugin.page

import org.objectweb.asm.ClassVisitor
import org.objectweb.asm.MethodVisitor
import org.objectweb.asm.Opcodes

/**
 * Created by znh on 09/21/2023
 * <p>
 * Description: ASM操作Class
 */
class PageClassVisitor(classVisitor: ClassVisitor?) : ClassVisitor(Opcodes.ASM9, classVisitor) {

    private var className: String? = null

    override fun visit(
        version: Int,
        access: Int,
        className: String?,
        signature: String?,
        superName: String?,
        interfaces: Array<out String>?
    ) {
        super.visit(version, access, className, signature, superName, interfaces)
        this.className = className
    }

    override fun visitMethod(
        access: Int,
        methodName: String?,
        descriptor: String?,
        signature: String?,
        exceptions: Array<out String>?
    ): MethodVisitor {
        val methodVisitor = cv.visitMethod(access, methodName, descriptor, signature, exceptions)
        if (access == 2 && methodName == "<init>" && descriptor == "()V") {
            return object : MethodVisitor(Opcodes.ASM9, methodVisitor) {
            }
        }
        if (methodName?.startsWith("onCreate") == true) {
            return PageMethodVisitor(methodVisitor, className, methodName)
        }
        if (methodName?.startsWith("onDestroy") == true) {
            return PageMethodVisitor(methodVisitor, className, methodName)
        }
        return methodVisitor
    }
}